Modern/ops/src/main/java/org/bonej/ops/ellipsoid/FindEllipsoidFromBoundaryPoints.java

```diff
@@ -0,0 +1,343 @@
+        Vector3d p = pnp.getVertex();
+        Vector3d np = pnp.getNormal();
+        Vector3d q = qnq.getVertex();
+        Vector3d nq = qnq.getNormal();
+
+        Matrix3d NRotation = new Matrix3d();
+
+        NRotation.m00 = -np.x() * nq.x();
+        NRotation.m01 = -np.x() * nq.y();
+        NRotation.m02 = -np.x() * nq.z();
+
+        NRotation.m10 = -np.y() * nq.x();
+        NRotation.m11 = -np.y() * nq.y();
+        NRotation.m12 = -np.y() * nq.z();
+
+        NRotation.m20 = -np.z() * nq.x();
+        NRotation.m21 = -np.z() * nq.y();
+        NRotation.m22 = -np.z() * nq.z();
+
+        Matrix3d NRotationTransposed = new Matrix3d(NRotation);
+        NRotationTransposed.transpose();
+        Matrix3d q2Rotation = new Matrix3d(NRotation);
+        q2Rotation.add(NRotationTransposed);
+        q2Rotation.scale(0.5);
+
+        double nqDotqHalf = nq.dot(q) / 2.0;
+        double npDotpHalf = np.dot(p) / 2.0;
+
+        Vector3d q2Translation = new Vector3d(np.x() * nqDotqHalf + nq.x() * npDotpHalf, np.y() * nqDotqHalf + nq.y() * npDotpHalf, np.z() * nqDotqHalf + nq.z() * npDotpHalf);
+
+        Matrix4d quadric2 = new Matrix4d(q2Rotation);
+        quadric2.setColumn(3, new Vector4d(q2Translation,1));
+        quadric2.setRow(3, new Vector4d(q2Translation,1));
+        quadric2.m33(-np.dot(p) * nq.dot(q));
+        return quadric2;
+    }
+
+
+    /**
+     * Calculates the elliptic cylinder denoted by Q_3 in the original paper
+     *
+     * @param pqr       three non-collinear points defining a plane
+     * @param ellipsoid ellipsoid through p,q, and r
+     * @return a 4x4 matrix representation of the elliptic cylinder defined by the intersection ellipse of pqr and qBar
+     */
+    private static Matrix4d getQ3(final List<Vector3d> pqr, Ellipsoid ellipsoid) {
+        final Vector3d p = new Vector3d(pqr.get(0));
+        final Vector3d q = new Vector3d(pqr.get(1));
+        final Vector3d r = new Vector3d(pqr.get(2));
+
+        Vector3d interiorPoint = new Vector3d(p);
+        interiorPoint.mul(-1);
+        interiorPoint.add(q);
+        interiorPoint.mul(0.5);
+        interiorPoint.add(p);
+
+        Vector3d planeNormal = new Vector3d();
+        q.mul(-1);
+        q.add(p);
+        r.mul(-1);
+        r.add(p);
+        q.cross(r, planeNormal);
+        List<Vector3d> ellipse = intersectionOp.calculate(ellipsoid, new ValuePair<>(interiorPoint, planeNormal));
+
+        Vector3dc a0 = ellipse.get(0);
+        Vector3dc a1 = ellipse.get(1);
+        Vector3dc a2 = ellipse.get(2);
+
+        Matrix4d quadric = new Matrix4d();
+
+        //axis and translation contributions calculated with sympy - see documentation
+        Matrix3d A1 = getAxisContribution(a1);
+        Matrix3d A2 = getAxisContribution(a2);
+        A2.add(A1);
+        quadric.set3x3(A2);
+
+        Vector3d translationVector1 = getTranslationContribution(a0, a1);
+        Vector3d translationVector2 = getTranslationContribution(a0, a2);
```

Comment:
We know from the types of the variables that they are vectors. No need to name them that.